vms/platformvm/utxo/locked.go

```diff
@@ -0,0 +1,691 @@
+			fee-amountBurned, // Amount we still need to burn
+			remainingValue,   // Amount available to burn
+		)
+		amountBurned += amountToBurn
+		remainingValue -= amountToBurn
+
+		// Stake any value that should be staked
+		amountToStake := math.Min(
+			amount-amountStaked, // Amount we still need to stake
+			remainingValue,      // Amount available to stake
+		)
+		amountStaked += amountToStake
+		remainingValue -= amountToStake
+
+		// Add the input to the consumed inputs
+		ins = append(ins, &avax.TransferableInput{
+			UTXOID: utxo.UTXOID,
+			Asset:  avax.Asset{ID: h.ctx.AVAXAssetID},
+			In:     in,
+		})
+
+		if amountToStake > 0 {
+			// Some of this input was put for staking
+			outs = append(outs, &avax.TransferableOutput{
+				Asset: avax.Asset{ID: h.ctx.AVAXAssetID},
+				Out: &lock.LockedOut{
+					LockIDs: lock.LockIDs{}.Lock(appliedLockState),
+					TransferableOut: &secp256k1fx.TransferOutput{
+						Amt: amountToStake,
+						OutputOwners: secp256k1fx.OutputOwners{
+							Locktime:  0,
+							Threshold: 1,
+							Addrs:     []ids.ShortID{changeAddr},
+						},
+					},
+				},
+			})
+		}
+
+		if remainingValue > 0 {
+			// This input had extra value, so some of it must be returned
+			outs = append(outs, &avax.TransferableOutput{
+				Asset: avax.Asset{ID: h.ctx.AVAXAssetID},
+				Out: &secp256k1fx.TransferOutput{
+					Amt: remainingValue,
+					OutputOwners: secp256k1fx.OutputOwners{
+						Locktime:  0,
+						Threshold: 1,
+						Addrs:     []ids.ShortID{changeAddr},
+					},
+				},
+			})
+		}
+
+		// Add the signers needed for this input to the set of signers
+		signers = append(signers, inSigners)
+	}
+
+	if amountBurned < fee || amountStaked < amount {
+		return nil, nil, nil, errNotEnoughBalance
+	}
+
+	avax.SortTransferableInputsWithSigners(ins, signers) // sort inputs and keys
+	avax.SortTransferableOutputs(outs, txs.Codec)        // sort outputs
+
+	return ins, outs, signers, nil
+}
+
+func (h *handler) Unlock(
+	state state.Chain,
+	lockTxIDs []ids.ID,
+	removedLockState lock.LockState,
+) (
+	[]*avax.TransferableInput, // inputs
+	[]*avax.TransferableOutput, // outputs
+	error,
+) {
+	if removedLockState != lock.LockStateBonded && removedLockState != lock.LockStateDeposited {
+		return nil, nil, errInvalidTargetLockState
+	}
+
+	lockTxAddresses := make(map[ids.ID]*ids.ShortSet, len(lockTxIDs))
+	for _, lockTxID := range lockTxIDs {
+		tx, s, err := state.GetTx(lockTxID)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to fetch lockedTx %s: %v", lockTxID, err)
+		}
+		if s != status.Committed {
+			return nil, nil, fmt.Errorf("%s is not a commited tx", lockTxID)
+		}
+
+		outputs := tx.Unsigned.Outputs()
+		for i, output := range outputs {
+			lockedOut, ok := output.Out.(*lock.LockedOut)
+			if !ok || !lockedOut.IsNewlyLockedWith(removedLockState) {
+				// we'r only intersed in outs locked by this tx
+				continue
+			}
+			innerOut, ok := lockedOut.TransferableOut.(*secp256k1fx.TransferOutput)
```

Comment:
Even we only deal here with secp256k1, it cann be made more flexible using the addressable interface